crates/pipeline_manager/src/db/mod.rs

```diff
@@ -994,6 +1058,136 @@ impl ProjectDB {
+        Ok(ret)
+    }
+
+    pub(crate) async fn create_or_replace_pipeline(
+        &self,
+        tenant_id: TenantId,
+        pipeline_name: &str,
+        program_name: &Option<String>,
+        pipeline_description: &str,
+        config: &RuntimeConfig,
+        connectors: &Option<Vec<AttachedConnector>>,
+    ) -> Result<(bool, PipelineId, Version), DBError> {
+        let mut manager = self.pool.get().await?;
+        let txn = manager.transaction().await?;
+        let pipeline = self
+            .get_pipeline_descr_by_name(tenant_id, pipeline_name, Some(&txn))
+            .await;
+        let res = match pipeline {
+            Ok(p) => Ok(self
+                .update_pipeline(
+                    tenant_id,
+                    p.pipeline_id,
+                    program_name,
+                    pipeline_name,
+                    pipeline_description,
+                    &Some(config.clone()),
+                    connectors,
+                    Some(&txn),
+                )
+                .await
+                .map(|v| (false, p.pipeline_id, v))?),
+            Err(DBError::UnknownPipelineName { pipeline_name }) => Ok(self
+                .new_pipeline(
+                    tenant_id,
+                    Uuid::now_v7(),
+                    program_name,
+                    &pipeline_name,
+                    pipeline_description,
+                    config,
+                    connectors,
+                    Some(&txn),
+                )
+                .await
+                .map(|(p, v)| (true, p, v))?),
+            Err(e) => Err(e),
+        }?;
+        txn.commit().await?;
+        Ok(res)
+    }
+
+    pub(crate) async fn create_or_replace_connector(
+        &self,
+        tenant_id: TenantId,
+        connector_name: &str,
+        description: &str,
+        config: &ConnectorConfig,
+    ) -> Result<(bool, ConnectorId), DBError> {
+        let mut manager = self.pool.get().await?;
+        let txn = manager.transaction().await?;
+        let connector = self
+            .get_connector_by_name(tenant_id, connector_name, Some(&txn))
+            .await;
+        match connector {
+            Ok(c) => Ok(self
+                .update_connector(
+                    tenant_id,
+                    c.connector_id,
+                    connector_name,
+                    description,
+                    &Some(config.clone()),
+                    Some(&txn),
+                )
+                .await
+                .map(|_| (false, c.connector_id))?),
+            Err(DBError::UnknownConnectorName { connector_name }) => Ok(self
+                .new_connector(
+                    tenant_id,
+                    Uuid::now_v7(),
+                    &connector_name,
+                    description,
+                    config,
+                    Some(&txn),
+                )
+                .await
+                .map(|c| (true, c))?),
+            Err(e) => Err(e),
+        }
```

Comment:
There should be a transaction commit here I think.

Reply:
Yes good catch!